[Chat] WebSocket 기반 채팅 기능 구현

yunjeooong opened the pull request:
Title
[Chat] WebSocket 기반 채팅 기능 구현
## Description
당근마켓과 유사하게 더쿠 장터 내에서 구매자와 판매자 간의 1:1 채팅 기능을 구현했습니다.
WebSocket을 사용하여 실시간 채팅이 가능하도록 하였으며, 채팅방 생성 및 조회 기능을 구현했습니다.


## Changes
- WebSocket 설정
STOMP 기반 실시간 양방향 통신 구현
/sub(구독), /pub(발행) 엔드포인트로 메시지 라우팅
WebConfig에서 CORS 설정 통합 관리

- 채팅방 도메인
채팅방 = 상품 + 구매자 + 판매자
읽지 않은 메시지 수 추적 (lastMessageId - lastReadMessageId)
ACTIVE/INACTIVE로 채팅방 상태 관리

- Repository & Service
사용자별(구매자/판매자) 채팅방 목록 조회
UUID 기반 roomId로 WebSocket 세션 관리
중복 채팅방 생성 방지

API
채팅방 생성: POST /api/chat/rooms
채팅방 목록: GET /api/chat/rooms?userId={userId}

추후 이어서
특정 채팅방 정보 조회
전체 안 읽은 채팅 개수 API 
api 개발 예정입니다.


**개선사항**
ChatRoomResponseDTO의 unreadCount 필드로 단순 카운트 방식으로 관리되고 있습니다 하지만 이러한  방식은 실시간 상태 관리의 어려움, 동시성 문제, 데이터 정합성 보장등등의 한계를 가지고 있어서 개선해야하는 상황입니다.

방법 1: 사용자의 마지막 접속 시간과 채팅 메시지 발송 시각 비교
방법 2: 안 읽은 메시지 수를 별도 컬럼/테이블에 저장
또는 더 나은 개선책으로 

협의 후 개선하겠습니다.

## Screenshots
## Ref

yunjeooong committed:
feat: WebSocket 채팅 기능을 위한 설정 추가

yunjeooong committed:
feat: 채팅방 관련 엔티티와 DTO 구현

- ChatRoom 엔티티 구현
  - 상품, 구매자, 판매자 정보 추가
  - 메시지 읽음 처리를 위한 필드 추가 (lastMessageId, buyerLastReadMessageId, sellerLastReadMessageId)
  - 채팅방 상태 관리를 위한 status 필드 추가

- DTO 클래스 구현
  - ChatRoomCreateRequestDTO: 채팅방 생성 요청 데이터
  - ChatRoomResponseDTO: 채팅방 정보 응답 데이터 (읽지 않은 메시지 수 포함)

yunjeooong committed:
feat: 채팅방 Repository 메서드 구현

yunjeooong committed:
feat: 채팅방 관리를 위한 ChatRoomService 구현

yunjeooong committed:
feat: 채팅방 Controller 구현

yunjeooong committed:
Merge branch 'develop' into feat/250113_chat

beetrbgus commented on src/main/java/com/ani/taku_backend/config/ChattingConfig.java:
```diff
@@ -0,0 +1,40 @@
+@EnableWebSocketMessageBroker
+public class ChattingConfig implements WebSocketMessageBrokerConfigurer {
+
+    private static final String DEVELOP_FRONT_ADDRESS = "http://localhost:3000";
```
이거 yml파일에 있는걸 @Value로 가져와 사용하는게 좋을 것 같습니다 -

beetrbgus commented on src/main/java/com/ani/taku_backend/config/ChattingConfig.java:
```diff
@@ -0,0 +1,40 @@
+
+@Configuration
+@EnableWebSocketMessageBroker
+public class ChattingConfig implements WebSocketMessageBrokerConfigurer {
```
클래스 이름이 채팅설정이 아닌 
알림 등 범용으로 사용할 거기 때문에 레디스 설정 이 맞을 것 같습니다

yunjeooong replied:
현재 RedisConfig는 Redis 연결 및 설정을 담당하고 있습니다!  하지만 Redis 설정과 WebSocket 설정은 서로 다른 책임을 가지고 있기 때문에 두 설정을 분리하는 것이 더 적절하다고 생각합니다. 대신 기존 클래스명인 ChattingConfig를 MessageBrokerConfig로 변경하는 것은 어떨가요?? 이렇게 하면 리뷰 내용 처럼  클래스가 채팅뿐만 아니라 알림 등 다양한 실시간 메시지 처리를 포함하는 의미를 담게 되어 더 좋은 설계가 설계가 될 것 같습니다~

beetrbgus replied:
아 Redis가 아니라 STOMP 설정이군요
말씀하신 MessageBrokerConfig가 적절할 것 같습니다.

beetrbgus commented on src/main/java/com/ani/taku_backend/config/WebConfig.java:
```diff
@@ -5,16 +5,35 @@
+    public void addCorsMappings(CorsRegistry registry) {
+        registry.addMapping("/**")
+                .allowedOrigins(DEVELOP_FRONT_ADDRESS)
+                .allowedMethods("GET", "POST", "PUT", "DELETE")
```
하드 코딩보다 HttpMethod 라는 enum 클래스가 있습니다.
물론 Http Method가 변할일은 거의 없지만 하드 코딩은 지양하는게 좋지 않을까라는 생각입니다!

yunjeooong committed:
refacor: 채팅 config관련  class 이름 변경

yunjeooong committed:
refactor: WebConfig CORS 설정 개선

- 하드코딩된 HTTP 메소드를 HttpMethod enum으로 변경
- 프론트엔드 URL을 설정 파일에서 주입받도록 수정
- 중복된 import 문 정리

yunjeooong committed:
Merge branch 'feat/250113_chat' of https://github.com/taku-project/taku-backend into feat/250113_chat
feat: 원격 저장소 변경사항 병합

nagul2 commented on src/main/java/com/ani/taku_backend/chatroom/model/constant/ChatRoomStatus.java:
common.enums 패키지에 StatusType으로 Enum이 구현되어있는데 기능이 중복인 것 같아서 구현하신 필드와 생성자들만 생성하여 사용하시는 것은 어떠신가용?

yunjeooong replied:
common.enums 패키지에 StatusType이라는 Enum이 이미 구현되어 있지만, 저는 이를 유저의 활성/비활성 상태를 나타내는 용도로 이해했습니다! 😊
반면, ChatRoomStatus는 채팅방의 상태를 나타내는 Enum으로, 다음과 같은 상태를 정의하고 있습니다:
	•	ACTIVE: 활성화된 채팅방
	•	현재 대화가 가능한 상태
	•	정상적으로 운영 중인 채팅방
	•	INACTIVE: 비활성화된 채팅방
	•	대화가 종료된 채팅방
	•	거래 완료, 채팅방 삭제 등으로 더 이상 사용되지 않는 상태

이 Enum은 사용자의 활성/비활성 상태가 아닌, 채팅방 자체의 상태를 관리하기 위한 목적으로 사용됩니다. 예를 들어:
	•	거래가 완료되면 채팅방 상태를 INACTIVE로 변경
	•	채팅방을 삭제하면 상태를 INACTIVE로 변경
	•	새로 생성된 채팅방은 ACTIVE 상태로 시작

이처럼 기능은 비슷해 보이지만, 의미와 목적이 다르기 때문에 별도로 관리하는 것이 적절하다고 판단했습니다.
다만, 기능적으로는 비슷한 만큼 다른 분들의 의견도 함께 들어보고 싶습니다! 🙌

nagul2 replied:
공통 패키지에 있다보니 어디에 종속되어있지않다고 생각해서 저는이미 장터 구현할때 status 필드를 StatusType로 구현하긴했어용 ㅎㅎ
크게 중요한 부분은 아닌것같아서 원하시는대로 하셔도 될 것 같습니다!
다만 장터 상태 필드는 해당 Enum으로 구현되어있어서 판매 api 구현하실때 참고가 필요하실 것 같습니다~

yunjeooong replied:
녭! 뭔가 확장성을 고려했을때는 뻗어 나갈 수 있는 방향이 다르다고 생각해서 일단은 분리했습니당. ex)  - ChatRoomStatus: ACTIVE, INACTIVE, SUSPENDED(임시정지), ARCHIVED(보관) 등,  - UserStatus: ACTIVE, INACTIVE, PENDING(승인대기), BLOCKED(차단) 등...
코드리뷰 감사드립니다!!!!😊

yunjeooong committed:
fix: 채팅 에러코드 복구

yunjeooong committed:
Merge branch 'develop' into feat/250113_chat
